tests/lib/eslint/flat-eslint.js

```diff
@@ -4501,17 +4501,10 @@ describe("FlatESLint", () => {
             });
         });
 
-
-        /*
-         * These tests fail due to a bug in fast-glob that doesn't allow
-         * negated patterns inside of ignores. These tests won't work until
-         * this bug is fixed:
-         * https://github.com/mrmlnc/fast-glob/issues/356
-         */
-        xdescribe("ignorePatterns can unignore '/node_modules/foo'.", () => {
+        describe("ignores can unignore '/node_modules/foo'.", () => {
 
             const { prepare, cleanup, getPath } = createCustomTeardown({
-                cwd: root,
+                cwd: `${root}-unignores`,
                 files: {
                     "eslint.config.js": `module.exports = {
                         ignores: ["!**/node_modules/foo/**"]
```

Comment:
Hmmm, are you sure about that? it seems like this test was indicating our behavior worked in a different way.

Generally we are treating patterns ending with star-str as file matchers and not directory matchers.

Reply:
Here is the current test, which verifies this behavior:
https://github.com/eslint/eslint/blob/main/tests/lib/eslint/eslint.js#L5413-L5460

Reply:
> Here is the current test, which verifies this behavior:
> https://github.com/eslint/eslint/blob/main/tests/lib/eslint/eslint.js#L5413-L5460

In the original test, the list of patterns is:

```
/**/node_modules/*
!/node_modules/foo
```

`node_modules/` doesn't match any patterns; `node_modules/foo/` matches `/**/node_modules/*` but also matches a subsequent negative pattern `!/node_modules/foo` so it's not ignored; `node_modules/foo/index.js` doesn't match any patterns.

In the flat eslint test, the list of patterns is:

```
**/node_modules/**
!**/node_modules/foo/**
```

`node_modules/` doesn't match any patterns; `node_modules/foo/` matches `**/node_modules/**` and doesn't match any negative patterns so `node_modules/foo/` is ignored and thus `node_modules/foo/index.js` should be ignored because it's in an ignored directory.

Reply:
> Generally we are treating patterns ending with star-str as file matchers and not directory matchers.

I think it would be confusing if trailing `/**` doesn't match subdirectories, as it is expected to match everything.

In my opinion, `foo/**` pattern should not match `foo/` directory, but it should match `foo/bar/` directory. This is how gitignore works. In minimatch, however, `foo/**` does match `foo/` directory, but that may be a bug (https://github.com/isaacs/minimatch/issues/179).

Reply:
Looks like that minimatch behavior is intentional.

I'll take some time to look at this again...it's such an edge case that I think there's room for disagreement on how to handle this. Because we are not explicitly guaranteeing to work the same way `gitignore` works, I think we can kind of chart our own course here.

Reply:
So interestingly, it appears that `ConfigArray` actually works the way you describe. I added some tests to verify:
https://github.com/humanwhocodes/config-array/pull/73

It looks like there must be a bug on the ESLint side of things, so I'm investigating that next.

Reply:
I'll take a look at https://github.com/humanwhocodes/config-array/pull/73.

Reply:
So it actually looks like it's a bug in `isDirectoryIgnored()` as opposed to `isFileIgnored()`.

Reply:
Fix is up, for when you have time:
https://github.com/humanwhocodes/config-array/pull/74